Migrate to Apollo Client 4

tiagojufr opened the pull request:
- Upgrade dev dependencies to support modern Node.js versions
- Emit ES6 code as Apollo Client 4 also does it and we can't have ES5 code extending an ES6 class (`ApolloLinkNetworkStatus extends ApolloLink`)

tiagojufr committed:
Migrate to Apollo Client 4

- Upgrade dev dependencies to support modern Node.js versions
- Emit ES6 code as Apollo Client 4 also does it and we can't have ES5 code extending an ES6 class (`ApolloLinkNetworkStatus extends ApolloLink`)

amannn commented:
Many thanks for looking into this and updating dependencies in general! 🙌 

@symn Do you by chance have time for a review?

symn commented:
Thanks for taking the time, i just have one question about the changed test

tiagojufr commented on src/__tests__/index-test.tsx:
```diff
@@ -139,7 +140,7 @@ it('can configure which operations to handle on a case-by-case basis', async ()
     <ConfiguredApolloProvider>
       <ConfiguredNetworkStatusReporter initialOptIn />
       <ConfiguredNetworkStatusReporter />
-      <DataFetcher initialSkip={false} />
```
That was the only way I could make that test pass. From what I tested in your example page in dev mode, everything works as it should when you toggle those check boxes that simulate the change of that prop. On the test environment, it doesn't seem to be behaving the same way and honestly I couldn't understand why 😕

tiagojufr replied:
pushed another commit to fix the warning shown when running `yarn lint`

zerocewl replied:
lgtm 👍

symn replied:
> Ok I'll do that but this needs to be a new major version because it doesn't work with Apollo client 3. I'll increase version to 6.0.0

of course 😅
>not sure when I'll have the time though as I'm on parental leave now

and congratulations - enjoy!

tiagojufr committed:
Fix warnings shown when launching in dev mode

- Fix type `export type`
- Bump some `eslint` related dependencies so `export type` does not break linting
- Pin `eslint-plugin-sort-destructure-keys` to a version compatible with eslint v7
- Add `import` plugin to `.eslintrc` to avoid `Definition for rule 'import/no-extraneous-dependencies' was not found` error
- Fix linting errors introduced by new dependency versions

tiagojufr committed:
Fix lint warnings

- Upgrade `eslint` to v8 to ensure compatibility with latest `typescript`
- Add `rimraf` to `devDependencies` to fix the `build` script on Windows

tiagojufr committed:
Remove unintentionally added `export`

symn reviewed (changes requested):
fix test, increase <del>patch</del> major version and add changelog please

zerocewl committed:
Adapt test to work with apollo client v4 update

zerocewl committed:
Update CHANGELOG.md for v6.0.0

zerocewl committed:
Bumped version to 6.0.0

tiagojufr committed:
Merge pull request #1 from zerocewl/master

Adapt test, increased major version and added changelog for apollo client 4

tiagojufr committed:
Fix changelog

tiagojufr commented:
@symn @amannn I think this is ready to be merged and released. Thank you very much @zerocewl for your PR!

symn reviewed (approved):
thank you for your participation @tiagojufr and @zerocewl

symn commented:
Closes https://github.com/molindo/react-apollo-network-status/issues/54
